# Part 1: Data cleaning

In [1]:
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings('ignore')


In [2]:
#creating path name var for the raw csv files

rounds2_path    = '/Users/manish/Documents/Projects/data_science/investment_strategy/data/raw_data/rounds2.csv'
companies_path = '/Users/manish/Documents/Projects/data_science/investment_strategy/data/raw_data/companies.csv'
mapping_path   = '/Users/manish/Documents/Projects/data_science/investment_strategy/data/raw_data/mapping.csv'

# Initial Analysis

## 1. Companies

In [3]:
#creating dataframes
companies = pd.read_csv(companies_path, encoding = 'ISO-8859-1')
companies.head()

permalink                    name  \
0                  /Organization/-Fame                   #fame   
1               /Organization/-Qounter                :Qounter   
2  /Organization/-The-One-Of-Them-Inc-  (THE) ONE of THEM,Inc.   
3                /Organization/0-6-Com                 0-6.com   
4       /Organization/004-Technologies        004 Technologies   

                        homepage_url  \
0                 http://livfame.com   
1             http://www.qounter.com   
2                http://oneofthem.jp   
3                 http://www.0-6.com   
4  http://004gmbh.de/en/004-interact   

                                       category_list     status country_code  \
0                                              Media  operating          IND   
1  Application Platforms|Real Time|Social Network...  operating          USA   
2                                  Apps|Games|Mobile  operating          NaN   
3                                        Curated Web  operating          CHN   
4                                           Software  operating          USA   

  state_code                 region           city  founded_at  
0         16                 Mumbai         Mumbai         NaN  
1         DE             DE - Other  Delaware City  04-09-2014  
2        NaN                    NaN            NaN         NaN  
3         22                Beijing        Beijing  01-01-2007  
4         IL  Springfield, Illinois      Champaign  01-01-2010

In [4]:
#basic info of companies file
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   permalink      66368 non-null  object
 1   name           66367 non-null  object
 2   homepage_url   61310 non-null  object
 3   category_list  63220 non-null  object
 4   status         66368 non-null  object
 5   country_code   59410 non-null  object
 6   state_code     57821 non-null  object
 7   region         58338 non-null  object
 8   city           58340 non-null  object
 9   founded_at     51147 non-null  object
dtypes: object(10)
memory usage: 5.1+ MB


In [5]:
#checking unique & top values and frequencies of the top values in all the columns
companies.describe()

permalink    name                homepage_url category_list  \
count                 66368   66367                       61310         63220   
unique                66368   66099                       61191         27296   
top     /Organization/-Fame  #NAME?  http://www.askforoffer.com      Software   
freq                      1       4                           5          3995   

           status country_code state_code       region           city  \
count       66368        59410      57821        58338          58340   
unique          4          137        311         1092           5111   
top     operating          USA         CA  SF Bay Area  San Francisco   
freq        53034        37601      12900         8804           3526   

        founded_at  
count        51147  
unique        3978  
top     01-01-2012  
freq          2730

In [6]:
#checking the missing values
companies.isnull().sum()

permalink            0
name                 1
homepage_url      5058
category_list     3148
status               0
country_code      6958
state_code        8547
region            8030
city              8028
founded_at       15221
dtype: int64

### Companies file analysis -
    
    Companies file has 66368 entries and 10 columns with various features

    1. permalink, home_url and name are three company specific identifiers and name column 
        has 1 missing value, with home_url with 5038 missing values.
    
    2. category_list gives the industry category the company belongs to.
    
    3. country_code, state_code, region_code and city are geography specific identifiers 
        and have some missing values.
    
    4. status has also some missing values along with founded_at.
    
    5. We will deal with each column one-by-one
    
    6. Lets get some initial analysis of rounds file.

## 2. Rounds

In [7]:
#creating dataframe
rounds2 = pd.read_csv(rounds2_path, encoding = 'ISO-8859-1')
rounds2.head()

company_permalink  \
0                  /organization/-fame   
1               /ORGANIZATION/-QOUNTER   
2               /organization/-qounter   
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-   
4                /organization/0-6-com   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B  05-01-2015         10000000.0  
1                  A  14-10-2014                NaN  
2                NaN  01-03-2014           700000.0  
3                  B  30-01-2014          3406878.0  
4                  A  19-03-2008          2000000.0

In [8]:
#basic info
rounds2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   company_permalink        114949 non-null  object 
 1   funding_round_permalink  114949 non-null  object 
 2   funding_round_type       114949 non-null  object 
 3   funding_round_code       31140 non-null   object 
 4   funded_at                114949 non-null  object 
 5   raised_amount_usd        94959 non-null   float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB


In [9]:
# general analysis of  numerical columns
rounds2.describe()

raised_amount_usd
count       9.495900e+04
mean        1.042687e+07
std         1.148212e+08
min         0.000000e+00
25%         3.225000e+05
50%         1.680511e+06
75%         7.000000e+06
max         2.127194e+10

In [10]:
#getting data for rest of the columns
rounds2.drop(columns = ['raised_amount_usd']).describe()

company_permalink  \
count                     114949   
unique                     90247   
top     /ORGANIZATION/SOLARFLARE   
freq                          10   

                                funding_round_permalink funding_round_type  \
count                                            114949             114949   
unique                                           114949                 14   
top     /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
freq                                                  1              55494   

       funding_round_code   funded_at  
count               31140      114949  
unique                  8        5033  
top                     A  01-01-2014  
freq                14747         779

In [11]:
rounds2.isnull().sum()

company_permalink              0
funding_round_permalink        0
funding_round_type             0
funding_round_code         83809
funded_at                      0
raised_amount_usd          19990
dtype: int64

### Rounds file analysis

    Rounds file has 114949 entries and 6 columns with various features
    
    1. company_permalink seems to be similar to that of the permalink column
    in companies file
    
    2. fundaing_round_permalink is another name identifier for the funding round
    
    3. funding round type gives us the idea of the type of funding done.
    
    4. funding_round_code is another feature with codes about the funding round. 
        1. It has 83809 missing values
    
    5. Raised_amount_usd gives the funding in USD.
        1. Max amount raised is 21.27 billion USD
        2. Min is 0 and average funding is 10 million USD
        3. The average is inflated due the high outliers.

## 3. Checking the common permalink column in both companies and rounds2 file

    * rounds2 has 90247 unique permalink 
    * companies has 66368 unique permalink entries
    * let us try to convert them to lower case and check

In [12]:
#this function returns converted column to lower case
def convert_to_lower(df):
    
    print("column to convert to lower case:\n")
    column = input('> ').strip().lower()
    
    df[column] = df[column].apply(lambda x: x.lower())

    return df
    

In [13]:
print("\n Convert rounds file's company_permalink column to lower case: \n")
rounds2 = convert_to_lower(rounds2)

column to convert to lower case:

> company_permalink


In [14]:
print("\nConvert companies file's permalink column to lower case:\n")
companies = convert_to_lower(companies)

column to convert to lower case:

> permalink


In [15]:
#checking the unique entries in both the data frame

print('rounds', len(rounds2.company_permalink.unique()))
print('companies', len(companies.permalink.unique()))

rounds 66370
companies 66368


    * This seems reasonable. BUt still there are two more unique entries in rounds2 df that 
      is not present in companies df

In [16]:
#Lets investigate further
rounds2.loc[~rounds2.company_permalink.isin(companies.permalink), :]['company_permalink']


29597                               /organization/e-cãŠbica
31863          /organization/energystone-games-灵石游戏
45176                  /organization/huizuche-com-惠ç§ÿ车
58473                /organization/magnet-tech-磁石科技
101036    /organization/tipcat-interactive-沙èˆÿ信息ç...
109969               /organization/weiche-tech-喂车科技
113839                   /organization/zengame-ç¦
游科技
Name: company_permalink, dtype: object

    * There seems to be some weird characters in the company_permalink in rounds file.
    * Lets check the csv file to see if these entries are actually the same as in dataframe
    * The csv file seems to have the above 7 entries in correct format.
    * This means there is an issue in decoding the file in python
    * The main reason that python is not able to decode is beacause of various languages this data 
      is collected from various countries. 
    * By searching some errors online, the following hack can be used

In [17]:
rounds2['company_permalink'] = rounds2.company_permalink.str.encode('utf-8').str.decode('ascii', 'ignore')
companies['permalink']       = companies.permalink.str.encode('utf-8').str.decode('ascii', 'ignore')

In [18]:
#lets if the hack worked
print('Uncommon permalink in rounds to that in companies:',
      rounds2.loc[~rounds2.company_permalink.isin(companies.permalink), :]['company_permalink'])

print('rounds', len(rounds2.company_permalink.unique()))
print('companies', len(companies.permalink.unique()))

Uncommon permalink in rounds to that in companies: Series([], Name: company_permalink, dtype: object)
rounds 66368
companies 66368


    * It seems the hack has worked.
    * We can put these clean files separately and not bother about encoding

In [19]:
#writing to clean csv files
#rounds2.to_csv("rounds2_clean.csv", index = False)
#companies.to_csv("companies_clean.csv", index = False)

### Lets move to data cleaning 2 notebook